toolkit/offchain/src/register.rs

```diff
@@ -1,17 +1,194 @@
+};
+use ogmios_client::{
+	query_ledger_state::{QueryLedgerState, QueryUtxoByUtxoId},
+	query_network::QueryNetwork,
+	transactions::Transactions,
+	types::OgmiosTx,
+	types::OgmiosUtxo,
+};
+use partner_chains_plutus_data::registered_candidates::{
+	block_producer_registration_to_plutus_data, RegisterValidatorDatum,
 };
-use ogmios_client::types::OgmiosUtxo;
-use partner_chains_plutus_data::registered_candidates::block_producer_registration_to_plutus_data;
-use sidechain_domain::BlockProducerRegistration;
+use sidechain_domain::{
+	BlockProducerRegistration, MainchainAddressHash, MainchainPrivateKey, MainchainPublicKey,
+	McTxHash, UtxoId, UtxoIndex,
+};
+
+pub trait Register {
+	#[allow(async_fn_in_trait)]
+	async fn register(
+		&self,
+		genesis_utxo: UtxoId,
+		block_producer_registration: &BlockProducerRegistration,
+		registration_utxo: UtxoId,
+		payment_signing_key: MainchainPrivateKey,
+	) -> Result<OgmiosTx, OffchainError>;
+}
+
+impl<T> Register for T
+where
+	T: QueryLedgerState + Transactions + QueryNetwork + QueryUtxoByUtxoId,
+{
+	async fn register(
+		&self,
+		genesis_utxo: UtxoId,
+		block_producer_registration: &BlockProducerRegistration,
+		registration_utxo: UtxoId,
+		payment_signing_key: MainchainPrivateKey,
+	) -> Result<OgmiosTx, OffchainError> {
+		run_register(
+			genesis_utxo,
+			block_producer_registration,
+			registration_utxo,
+			payment_signing_key,
+			self,
+			FixedDelayRetries::two_minutes(),
+		)
+		.await
+		.map_err(|e| OffchainError::InternalError(e.to_string()))
+	}
+}
+
+pub async fn run_register<
+	C: QueryLedgerState + QueryNetwork + QueryUtxoByUtxoId + Transactions,
+	A: AwaitTx,
+>(
+	genesis_utxo: UtxoId,
+	block_producer_registration: &BlockProducerRegistration,
+	registration_utxo: UtxoId,
+	payment_signing_key: MainchainPrivateKey,
+	ogmios_client: &C,
+	await_tx: A,
+) -> anyhow::Result<OgmiosTx> {
+	let ctx = TransactionContext::for_payment_key(payment_signing_key.0, ogmios_client).await?;
+	let own_pkh = ed25519_key_hash_to_mainchain_address_hash(ctx.payment_key_hash());
+	let validator = crate::scripts_data::registered_candidates_scripts(genesis_utxo)?;
+	let validator_address = validator.address_bech32(ctx.network)?;
+	let registration_utxo = ctx
+		.payment_key_utxos
+		.iter()
+		.find(|u| u.to_domain() == registration_utxo)
+		.ok_or(anyhow!("registration utxo not found at payment address"))?;
+	let all_registration_utxos = ogmios_client.query_utxos(&[validator_address]).await?;
+	let own_registrations = get_own_registrations(
+		own_pkh,
+		block_producer_registration.stake_ownership.pub_key.clone(),
+		&all_registration_utxos,
+	)?;
+	let own_registration_utxos = own_registrations.iter().map(|r| r.0.clone()).collect::<Vec<_>>();
+
+	if own_registrations.iter().any(|(_, existing_registration)| {
+		block_producer_registration.matches_keys(existing_registration)
+	}) {
+		return Err(anyhow!("BlockProducer with given set of keys is already registered"));
```

Comment:
This code prevents users fixing their existing registration. Imagine they have given wrong signatures. Instead of matching keys, everything should be checked.

In case of perfect match, return None instead of error. This way you have nice idempotent API.
User wants particular registration to be on chain. So if he runs same command again, it will be Ok(None).

Reply:
I copied the logic from the existing offchain code. You do make a good point though, I will fix this.